### PR TITLE
Add support for nullable analysis in interpolated string handler constructors

### PR DESCRIPTION
**Pull request by 333fred**
Implements most of https://github.com/dotnet/roslyn/issues/54583. Arguments with slots are treated like we do receivers: as though the previous argument value has already been loaded, and therefore postcondition attributes do not flow back to the original argument position (the effect of this can be seen in StringInterpolation_14 and _20), but future references to those locations will see the effects of the attribute.

I additionally looked at updating region analysis as part of this PR to see if I could share data between these passes, but I don't believe we can do much sharing here. There's too much nullable-specific with finding the current state of the expression being used for the mapping to work.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:64382]: I feel silly for asking at this stage but.. I am wondering what is supposed to happen when the user sees this in intellisense? Are most APIs going to be overloaded on 'string' and 'CustomHandler' in practice so the user will see the string overload and say "ok, I pretty much know what to provide here"?

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:64382]: Likely most will overload, but nothing special will be shown for custom handlers.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3284]: It feels like whenever we might have a not-null receiver, we pass a NotNull flow state here. Is it necessary for this to be its own parameter?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3013]: >LeastOverriddenMember [](http://example.com/codeflow?start=152&length=21)

It is not obvious why we have to pass the ```LeastOverriddenMember``` here. If it is a requirement specific to binding an interpolated string handler, perhaps ```BindInterpolatedStringHandlerInMemberCall``` should do that itself rather than relying on the caller to supply the right value.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:969]: >when receiver != null [](http://example.com/codeflow?start=86&length=21)

It looks like the switch above over the same ```argumentIndex``` assumes the ```receiver``` is not null for ```case BoundInterpolatedStringArgumentPlaceholder.InstanceParameter```. Here we are guarding against null in the same situation. Is there a reason for the inconsistency?

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:971]: >valSafeToEscapeScope [](http://example.com/codeflow?start=24&length=20)

Is this change related to the stated goal of the PR, i.e. support for nullable analysis? Is there a change in behavior here?

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:970]: >placeholderSyntax = unconvertedString.Syntax; [](http://example.com/codeflow?start=24&length=45)

Is this assignment necessary.

**Comment by AlekseyTs** [comment]
@333fred It looks like there are legitimate test failures.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:4253]: >_ = TryGetPlaceholderSlot(expressionWithoutConversion, out slot); [](http://example.com/codeflow?start=16&length=65)

Should MakeSlot take care of this side check instead?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:4173]: >if (TryGetPlaceholderSlot(operand, out slot) && PossiblyNullableType(operand.Type)) [](http://example.com/codeflow?start=28&length=83)

It feels like we wouldn't need this special case if MakeSlot took care of the side lookup.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5643]: If I am correctly interpreting the intent here, I think the following condition would be easier to follow:
```possibleInterpolationData is not null && possibleInterpolationData.ArgumentPlaceholders.Length > (possibleInterpolationData.HasTrailingHandlerValidityParameter ? 1 : 0)```. Then the block can keep using ```possibleInterpolationData```.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5654]: >State[slot] = receiverState; [](http://example.com/codeflow?start=32&length=28)

It is not clear to me why overwriting the state for the real receiver's slot is the right thing to do.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5660]: >State[slot] [](http://example.com/codeflow?start=36&length=11)

Same comment here.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3366]: >BoundExpression [](http://example.com/codeflow?start=43&length=15)

It looks like this function is only used for BoundInterpolatedStringArgumentPlaceholder. I think we can benefit from making the parameter type stronger.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3368]: >_placeholderLocalsOpt [](http://example.com/codeflow?start=16&length=21)

I think the implementation can be simpler if we change ```getInterpolatedStringHandlerPlaceholderSlot``` to always add an entry into ```_interpolatedStringArgumentPlaceholderAliases```. This way we wouldn't need to look in two places.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3366]: >TryGetPlaceholderSlot [](http://example.com/codeflow?start=21&length=21)

The name feels somewhat confusing given the name of the previous methods. It feels like that apply different meaning to the "Placeholder" word. Perhaps this name should be more specific.

**Comment by AlekseyTs** [comment]
> Arguments with slots are treated like we do receivers: as though the previous argument value has already been loaded, and therefore postcondition attributes do not flow back to the original argument position

Is this something that should be approved by LDM? Also, do we plan to update nullable specification?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:9713]: >receiverOpt = null; [](http://example.com/codeflow?start=16&length=19)

It is not clear to me why we start pretending that there is no receiver. Perhaps, this allows us to get what we want (BTW, what is it) from ```VisitArgumentsEvaluate``` call below today. But we might get in trouble if we start using this value for something else.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5617]: >receiverState [](http://example.com/codeflow?start=30&length=13)

It looks like we always pass one of the two values:
- default
- NotNull.

More over, default is passed only when we explicitly pass ```nulll``` for the receiver. What is the advantage of using ```default``` state vs. NotNull state in cases like that? Perhaps we don't need to pass the ```receiverState``` explicitly at all?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:971]: > Is this change related to the stated goal of the PR, i.e. support for nullable analysis? Is there a change in behavior here?

Yes, it's related, though this itself shouldn't cause a behavior change. I needed to pass through additional context from the receiver for the correct syntax node to report nullability warnings, and whether the receiver has a null suppression on it. I decided to simplify the existing receiver data to just pass the receiver itself.

**Comment by 333fred** [comment]
Converting this to a draft PR while I'm out on vacation.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:3013]: I'll reduce the symbol to just a bool for whether it requires an instance receiver.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3366]: This function is being removed.

**Comment by 333fred** [comment]
> Also, do we plan to update nullable specification?

I'll update the specification with the limitations after this is merged.

---
In reply to: [974418424](https://github.com/dotnet/roslyn/pull/57780#issuecomment-974418424) [](http://example.com/codeflow?ancestors=974418424,974418424)

**Comment by 333fred** [comment]
@AlekseyTs addressed your feedback and did a bit of rearranging. I did have to merge from `main` to pull in the placeholder work you did, so the diff isn't the cleanest between 2 and 9, unfortunately.

---
In reply to: [974438542](https://github.com/dotnet/roslyn/pull/57780#issuecomment-974438542) [](http://example.com/codeflow?ancestors=974438542,974438542)

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:255]: >throw ExceptionUtilities.UnexpectedValue(e.Kind), [](http://example.com/codeflow?start=21&length=49)

```!throwOnMissing ? default : throw ExceptionUtilities.UnexpectedValue(e.Kind)``` and get rid of the previous item in the switch?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs:255]: >e.Kind [](http://example.com/codeflow?start=62&length=6)

Complaining about unexpected node for ```BoundBinaryOperator``` or ```BoundInterpolatedString``` might be confusing.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5601]: >argument [](http://example.com/codeflow?start=57&length=8)

It looks like we cold keep the original code for the body of this function.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:5679]: >conversionResultsBuilder.Count == argumentIndex) [](http://example.com/codeflow?start=61&length=48)

Could we infer the index from the count then?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:7000]: >info.PreviousArgumentConverionResults.Count == info.ArgumentIndex [](http://example.com/codeflow?start=82&length=65)

Could we infer the index from the count then?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 9), it looks like there are merge conflicts to resolve.

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [987109770](https://github.com/dotnet/roslyn/pull/57780#issuecomment-987109770) [](http://example.com/codeflow?ancestors=987109770,987109770)

**Comment by AlekseyTs** [comment]
@333fred It looks like there are legitimate test failures

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:9996]: >SetNotNullResult [](http://example.com/codeflow?start=12&length=16)

VisitPlaceholderWithReplacement?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 11).

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: I am unsure how this got reverted, must have been something with the merge?

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback, thanks.

---
In reply to: [987441294](https://github.com/dotnet/roslyn/pull/57780#issuecomment-987441294) [](http://example.com/codeflow?ancestors=987441294,987441294)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:7557]: >_resultForPlaceholdersOpt.Remove [](http://example.com/codeflow?start=28&length=32)

```RemovePlaceholderReplacement```? And remove the assert before the loop?

**Review by AlekseyTs — DISMISSED** [review]
LGTM (commit 12), modulo the small refactoring suggestion around an explicit use of ```_resultForPlaceholdersOpt``` field.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler @RikkiGibson for a second review.

**Comment by 333fred** [comment]
@AlekseyTs, a test I added in this PR revealed a bug in the IOperation implementation, please take a look at the latest commit.

File: src/Compilers/Test/Core/Compilation/TestOperationVisitor.cs
Comment (333fred) [reply src/Compilers/Test/Core/Compilation/TestOperationVisitor.cs:1548]: Because, in the case of interpolated string handlers, the declaration expression is actually captured. See the graph in the test case I added, which would fail here because the declaration expression goes through here.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:1853]: >CloneOperation [](http://example.com/codeflow?start=55&length=14)

Is the clone necessary?

Comment (333fred) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:1853]: Yes. The expression already has a `Parent` and we can't reset it.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:1805]: It feels like we might spill too much. I.e. spill `DeclarationExpression`s that do not belong to the immediate explicit call. We probably should limit this exception by ```_currentInterpolatedStringHandlerArgumentContext.StartingStackDepth```, but then we also might need to drop the optimization based on ```_startSpillingAt```.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 15)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:1805]: On the second thought, this optimization will come at the cost of figuring out the right spill region for the spilling "out of order". Not worth the risk and complexity, I think.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:6415]: >spillDeclarationExpressions: true [](http://example.com/codeflow?start=27&length=33)

This doesn't feel sufficient. This parameter will make the difference only for entries that haven't been spilled already. And there is no guarantee that that will be the case. For example, if an argument following a ```DeclarationExpression``` has a control flow, it will case spilling for the ```DeclarationExpression``` and after that this call won't do anything about it. It should be easy to come up with a test case that exposes this issue.
Some options to address this that come to mind:
- Unconditionally unwrap ```DeclarationExpression```s during spilling. If we can run into tuple-like ```DeclarationExpression```s, perhaps we can skip those.
- Preemptively unwrap ```DeclarationExpression```s when we push arguments to the stack.
- Handle spilling of ```DeclarationExpression```out of order on demand once we realize that we need to use it multiple times due to the interpolation.
We can discuss our options in more details offline.

**Comment by 333fred** [comment]
@AlekseyTs I implemented the change we talked about offline. Please take another look when you get the chance.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:1805]: >bool spillDeclarationExpressions = false [](http://example.com/codeflow?start=36&length=40)

Is the parameter still necessary.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringOperation.cs:5944]: >); [](http://example.com/codeflow?start=48&length=2)

Consider adding a test with declaration expression between two interpolations and also a test demonstrating that declaration expressions at the end survive.

**Comment by AlekseyTs** [comment]
>         private static readonly Func<IArgumentOperation, IOperation> UnwrapArgument = UnwrapArgumentDoNotCaptureDirectly;

Is this delegate used?

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:2092 in 733946f. [](commit_id = 733946f06683acf102fa7505efd453721f59603f, deletion_comment = False)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:2085]: >SpillEvalStack(spillDeclarationExpressions: true); [](http://example.com/codeflow?start=20&length=50)

Instead of doing this, I would simply conditionally unwrap declaration expressions before calling ```VisitRequired``` above.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 16).

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:2085]: Leaving the logic in SpillEvalStack feels fragile because it will continue to be a trap for user that might think that passing ```true``` actually provides guarantees on its own. Another shortcoming is that this call is going to unwrap declaration expressions that do not belong to this argument list. For example, arguments of an enclosing call. Please add a test case for this scenario.

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [991353417](https://github.com/dotnet/roslyn/pull/57780#issuecomment-991353417) [](http://example.com/codeflow?ancestors=991353417,991353417)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:2069]: >complexity [](http://example.com/codeflow?start=38&length=10)

"complexity" used twice. Consider rephrasing.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 17), assuming CI is passing.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler @RikkiGibson for a second review.

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**
